shell/platform/darwin/macos/framework/Source/FlutterOpenGLRenderer.mm

```diff
@@ -0,0 +1,165 @@
+  return [engine.openGLRenderer makeCurrent];
+}
+
+static bool OnClearCurrent(FlutterEngine* engine) {
+  return [engine.openGLRenderer clearCurrent];
+}
+
+static bool OnPresent(FlutterEngine* engine) {
+  return [engine.openGLRenderer present];
+}
+
+static uint32_t OnFBO(FlutterEngine* engine, const FlutterFrameInfo* info) {
+  return [engine.openGLRenderer getFBO:info];
+}
+
+static bool OnMakeResourceCurrent(FlutterEngine* engine) {
+  return [engine.openGLRenderer makeResourceCurrent];
+}
+
+static bool OnAcquireExternalTexture(FlutterEngine* engine,
+                                     int64_t textureIdentifier,
+                                     size_t width,
+                                     size_t height,
+                                     FlutterOpenGLTexture* openGlTexture) {
+  return [engine.openGLRenderer populateTextureWithIdentifier:textureIdentifier
+                                                openGLTexture:openGlTexture];
+}
+
+#pragma mark - FlutterOpenGLRenderer implementation.
+
+@implementation FlutterOpenGLRenderer {
+  // The embedding-API-level engine object.
+  FLUTTER_API_SYMBOL(FlutterEngine) _engine;
+
+  FlutterView* _flutterView;
+
+  // The context that is owned by the currently displayed FlutterView. This is stashed
+  // in the renderer so that the view doesn't need to be accessed from a background thread.
+  NSOpenGLContext* _openGLContext;
+
+  // The context provided to the Flutter engine for resource loading.
+  NSOpenGLContext* _resourceContext;
+
+  // A mapping of textureID to internal FlutterExternalTextureGL adapter.
+  NSMutableDictionary<NSNumber*, FlutterExternalTextureGL*>* _textures;
+
+  FlutterEngineProcTable _embedderAPI;
+}
+
+- (instancetype)initWithFlutterEngine:(FLUTTER_API_SYMBOL(FlutterEngine))engine {
+  self = [super init];
+  if (self) {
+    _engine = engine;
+    _textures = [[NSMutableDictionary alloc] init];
+  }
+  return self;
+}
+
+- (void)attachToFlutterView:(FlutterView*)view {
+  _flutterView = view;
+  _openGLContext = view.openGLContext;
+}
+
+- (bool)makeCurrent {
+  if (!_openGLContext) {
+    return false;
+  }
+  [_openGLContext makeCurrentContext];
+  return true;
+}
+
+- (bool)clearCurrent {
+  [NSOpenGLContext clearCurrentContext];
+  return true;
+}
+
+- (bool)present {
+  if (!_openGLContext) {
+    return false;
+  }
+  [_flutterView present];
+  return true;
+}
+
+- (uint32_t)getFBO:(const FlutterFrameInfo*)info {
+  CGSize size = CGSizeMake(info->size.width, info->size.height);
+  return [_flutterView frameBufferIDForSize:size];
+}
+
+- (NSOpenGLContext*)resourceContext {
+  if (!_resourceContext) {
+    NSOpenGLPixelFormatAttribute attributes[] = {
+        NSOpenGLPFAColorSize, 24, NSOpenGLPFAAlphaSize, 8, 0,
+    };
+    NSOpenGLPixelFormat* pixelFormat = [[NSOpenGLPixelFormat alloc] initWithAttributes:attributes];
+    _resourceContext = [[NSOpenGLContext alloc] initWithFormat:pixelFormat shareContext:nil];
```

Comment:
Just for my clarification: How is the `shareContext` `nil` here? Are we using creating this context first and then using it as the prototype for the main context?

Reply:
Yup, exactly: https://github.com/flutter/engine/blob/master/shell/platform/darwin/macos/framework/Source/FlutterViewController.mm#L247